### PR TITLE
[SPARK-34353][SQL] CollectLimitExec avoid shuffle if input rdd has 0/1 partition

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
if child rdd has only one partition or zero partition, skip the shuffle

### Why are the changes needed?
skip shuffle if possible

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
existing testsuites

### CODE REVIEW

**Comment by zhengruifeng** [comment]
> We may have more operators that adding shuffle in the doExecute method instead of the planner

@cloud-fan shuffle only is directly added in the `doExecute` method of `TakeOrderedAndProjectExec`, `CollectLimitExec` and `ShuffleExchangeExec`

**Comment by zhengruifeng** [comment]
related to https://github.com/apache/spark/pull/31409

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
cc @viirya @maropu

**Review by HyukjinKwon — APPROVED** [review]

**Review by viirya — APPROVED** [review]

**Comment by cloud-fan** [comment]
retest this please

**Comment by viirya** [comment]
`org.apache.spark.sql.CachedTableSuite.SPARK-34269: cache lookup with ORDER BY / LIMIT clause` failed more than one time. But seems it passed without this change?

**Comment by maropu** [comment]
The fix itself looks fine.

**Comment by zhengruifeng** [comment]
retest this please

**Comment by zhengruifeng** [comment]
```
scala> spark.sql("CREATE TABLE t (key bigint, value string) USING parquet")
res0: org.apache.spark.sql.DataFrame = []

scala> spark.sql("SELECT COUNT(*) FROM t").head
res1: org.apache.spark.sql.Row = [0]

scala> val t = spark.table("t")
t: org.apache.spark.sql.DataFrame = [key: bigint, value: string]

scala> t.count
res2: Long = 0

scala> t.rdd.getNumPartitions
res3: Int = 0

scala> 

scala> spark.sql(s"CACHE TABLE v1 AS SELECT * FROM t LIMIT 10")
java.util.NoSuchElementException: next on empty iterator
  at scala.collection.Iterator$$anon$2.next(Iterator.scala:41)
  at scala.collection.Iterator$$anon$2.next(Iterator.scala:39)
  at scala.collection.IndexedSeqLike$Elements.next(IndexedSeqLike.scala:63)
  at scala.collection.IterableLike.head(IterableLike.scala:109)
  at scala.collection.IterableLike.head$(IterableLike.scala:108)
  at scala.collection.mutable.ArrayOps$ofRef.scala$collection$IndexedSeqOptimized$$super$head(ArrayOps.scala:198)
  at scala.collection.IndexedSeqOptimized.head(IndexedSeqOptimized.scala:129)
  at scala.collection.IndexedSeqOptimized.head$(IndexedSeqOptimized.scala:129)
  at scala.collection.mutable.ArrayOps$ofRef.head(ArrayOps.scala:198)
  at org.apache.spark.sql.Dataset.$anonfun$count$1(Dataset.scala:3024)
  at org.apache.spark.sql.Dataset.$anonfun$count$1$adapted(Dataset.scala:3023)
  at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:3705)
  at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:103)
  at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)
  at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:90)
  at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:772)
  at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
  at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3703)
  at org.apache.spark.sql.Dataset.count(Dataset.scala:3023)
  at org.apache.spark.sql.execution.datasources.v2.BaseCacheTableExec.run(CacheTableExec.scala:65)
  at org.apache.spark.sql.execution.datasources.v2.BaseCacheTableExec.run$(CacheTableExec.scala:41)
  at org.apache.spark.sql.execution.datasources.v2.CacheTableAsSelectExec.run(CacheTableExec.scala:88)
  at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.result$lzycompute(V2CommandExec.scala:42)
  at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.result(V2CommandExec.scala:42)
  at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.executeCollect(V2CommandExec.scala:48)
  at org.apache.spark.sql.Dataset.$anonfun$logicalPlan$1(Dataset.scala:228)
  at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:3705)
  at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:103)
  at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)
  at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:90)
  at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:772)
  at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
  at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3703)
  at org.apache.spark.sql.Dataset.<init>(Dataset.scala:228)
  at org.apache.spark.sql.Dataset$.$anonfun$ofRows$2(Dataset.scala:99)
  at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:772)
  at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:96)
  at org.apache.spark.sql.SparkSession.$anonfun$sql$1(SparkSession.scala:615)
  at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:772)
  at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:610)
  ... 47 elided

scala> spark.sql("SELECT COUNT(*) FROM v1").head
java.util.NoSuchElementException: next on empty iterator
  at scala.collection.Iterator$$anon$2.next(Iterator.scala:41)
  at scala.collection.Iterator$$anon$2.next(Iterator.scala:39)
  at scala.collection.IndexedSeqLike$Elements.next(IndexedSeqLike.scala:63)
  at scala.collection.IterableLike.head(IterableLike.scala:109)
  at scala.collection.IterableLike.head$(IterableLike.scala:108)
  at scala.collection.mutable.ArrayOps$ofRef.scala$collection$IndexedSeqOptimized$$super$head(ArrayOps.scala:198)
  at scala.collection.IndexedSeqOptimized.head(IndexedSeqOptimized.scala:129)
  at scala.collection.IndexedSeqOptimized.head$(IndexedSeqOptimized.scala:129)
  at scala.collection.mutable.ArrayOps$ofRef.head(ArrayOps.scala:198)
  at org.apache.spark.sql.Dataset.head(Dataset.scala:2747)
  ... 47 elided

scala> val v1 = spark.table("v1")
v1: org.apache.spark.sql.DataFrame = [key: bigint, value: string]

scala> v1.count
java.util.NoSuchElementException: next on empty iterator
  at scala.collection.Iterator$$anon$2.next(Iterator.scala:41)
  at scala.collection.Iterator$$anon$2.next(Iterator.scala:39)
  at scala.collection.IndexedSeqLike$Elements.next(IndexedSeqLike.scala:63)
  at scala.collection.IterableLike.head(IterableLike.scala:109)
  at scala.collection.IterableLike.head$(IterableLike.scala:108)
  at scala.collection.mutable.ArrayOps$ofRef.scala$collection$IndexedSeqOptimized$$super$head(ArrayOps.scala:198)
  at scala.collection.IndexedSeqOptimized.head(IndexedSeqOptimized.scala:129)
  at scala.collection.IndexedSeqOptimized.head$(IndexedSeqOptimized.scala:129)
  at scala.collection.mutable.ArrayOps$ofRef.head(ArrayOps.scala:198)
  at org.apache.spark.sql.Dataset.$anonfun$count$1(Dataset.scala:3024)
  at org.apache.spark.sql.Dataset.$anonfun$count$1$adapted(Dataset.scala:3023)
  at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:3705)
  at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$5(SQLExecution.scala:103)
  at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:163)
  at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:90)
  at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:772)
  at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:64)
  at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3703)
  at org.apache.spark.sql.Dataset.count(Dataset.scala:3023)
  ... 47 elided

scala> v1.rdd.getNumPartitions
res7: Int = 0

scala> v1.repartition(3).count
res8: Long = 0
```

it seem that `count` method does not work if the number of partitions of a **View** is zero.

`childRDD` may have no partition, then the number output partition is zero; 
while in existing impl, the `ShuffledRowRDD` will make sure `SinglePartition`, so that this issue will not be triggered.

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 1]
```diff
@@ -200,7 +205,7 @@ case class TakeOrderedAndProjectExec(
   protected override def doExecute(): RDD[InternalRow] = {
     val ord = new LazilyGeneratedOrdering(sortOrder, child.output)
     val childRDD = child.execute()
-    val singlePartitionRDD = if (childRDD.getNumPartitions > 1) {
+    val singlePartitionRDD = if (childRDD.getNumPartitions != 1) {
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala:208]: We don't have any test for the case `childRDD.getNumPartitions == 0`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 2]
```diff
@@ -52,16 +52,21 @@ case class CollectLimitExec(limit: Int, child: SparkPlan) extends LimitExec {
...
-      ShuffleExchangeExec.prepareShuffleDependency(
-        locallyLimited,
-        child.output,
-        SinglePartition,
-        serializer,
-        writeMetrics),
-      readMetrics)
-    shuffled.mapPartitionsInternal(_.take(limit))
+    val childRDD = child.execute()
+    val singlePartitionRDD = if (childRDD.getNumPartitions != 1) {
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala:56]: nit: could you leave some comments about why we cannot strip the shuffle if  `childRDD.getNumPartitions == 0`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala (hunk 1 above)
Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: There is no separate testsuite for `CollectLimitExec`, I added case `childRDD.getNumPartitions == 0` and case `childRDD.getNumPartitions == 1` in `TakeOrderedAndProjectSuite`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala (hunk 2 above)
Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala]: refer to [CoalesceExec.doExecute](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/basicPhysicalOperators.scala#L715), I update the fix to directly return empty RDD with single partition if `childRDD.getNumPartitions == 0`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 3]
```diff
@@ -200,28 +210,32 @@ case class TakeOrderedAndProjectExec(
...
-      new ShuffledRowRDD(
-        ShuffleExchangeExec.prepareShuffleDependency(
-          localTopK,
-          child.output,
-          SinglePartition,
-          serializer,
-          writeMetrics),
-        readMetrics)
+    if (childRDD.getNumPartitions == 0) {
+      sparkContext.parallelize(Array.empty[InternalRow], 1)
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala:214]: nit: how about using `ParallelCollectionRDD` directly?

Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala:214]: ok, I will update it.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by maropu** [comment]
Looks fine. Could you update the PR title and description, too? This PR is not only for a single partition case now.

**Review by maropu — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala [hunk 4]
```diff
@@ -52,16 +53,25 @@ case class CollectLimitExec(limit: Int, child: SparkPlan) extends LimitExec {
...
-        locallyLimited,
-        child.output,
-        SinglePartition,
-        serializer,
-        writeMetrics),
-      readMetrics)
-    shuffled.mapPartitionsInternal(_.take(limit))
+    val childRDD = child.execute()
+    if (childRDD.getNumPartitions == 0) {
+      new ParallelCollectionRDD(sparkContext, Seq.empty[InternalRow], 1, Map.empty)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala:58]: EmptyRDD?

Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala:58]: kindly ping @zhenglaizhang

Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala:58]: @viirya @maropu  we can not set the number of partitions in an `EmptyRDD`, it always has zero partitions.

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala:58]: Isn't the `childRDD` zero partitions?

Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala:58]: `new ParallelCollectionRDD(sparkContext, Seq.empty[InternalRow], 1, Map.empty)` is a empty rdd with single partition

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala:58]: Oh I see. I am not sure if `CollectLimitExec` must be single partition, but this looks minor. I'm okay with `ParallelCollectionRDD`. Thanks.

**Review by viirya — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala (hunk 4 above)
Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala:58]: @viirya the `outputPartitioning` of `CollectLimitExec` is  `SinglePartition`, so the output rdd should have single partition.

Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala:58]: or we can use `EmptyRDDWithPartitions` defined in `CoalesceExec`?
maybe we can make `EmptyRDD` support number of partition, so we can use it in both `CollectLimitExec` and `CoalesceExec`. But I am not sure whether we should do this.

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala:58]: > @viirya the `outputPartitioning` of `CollectLimitExec` is `SinglePartition`, so the output rdd should have single partition.

Oh, my previous comment was confusing. I mean I am not sure if its `outputPartitioning` must be ` SinglePartition`.

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/limit.scala:58]: > or we can use `EmptyRDDWithPartitions` defined in `CoalesceExec`?
> maybe we can make `EmptyRDD` support number of partition, so we can use it in both `CollectLimitExec` and `CoalesceExec`. But I am not sure whether we should do this.

I think it sounds over-engineering. At least for now I don't think we need it.

**Comment by maropu** [comment]
retest this please

**Comment by maropu** [comment]
run one more tests on the latest master before merging this.

**Comment by maropu** [comment]
retest this please

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by maropu** [comment]
Thanks! Merged to master.

**Comment by zhengruifeng** [comment]
Thank you so much !

**Pull request closed without merging**